handout/testing/Unit Testing Components.md

```diff
@@ -0,0 +1,240 @@
+
+## Injecting Dependencies and DOM Changes
+
+In the previous example the class we were testing `MessageComponent` did not have any injected dependencies. In Angular 2, components will often rely on services and other components in order to function, and these will be injected into the constructor of the class. When testing these components we have to inject the dependencies ourselves. Since this is an Angular specific routine, there are no pure Jasmine functions used to do this. Angular provides a list of functions in `angular2/testing` that allow us to to fully test our components. Lets take a look at a basic component:
+
+*quote.component.ts*
+
+``` typescript
+import {QuoteService} from './quote.service';
+import {Component} from 'angular2/core';
+
+@Component({
+	selector: 'my-quote',
+    template: '<h3>Random Quote</h3> <div>{{quote}}</div>'
+});
+
+export class QuoteComponent {
+  quote: string;
+  
+  constructor (private quoteService: QuoteService){}
+  
+  getQuote() {
+  	this.quoteService.getQuote().then((quote) => {
+  		this.quote = quote;
+	});
+   }
+}
+```
+
+This component relies on the `QuoteService` to get a random quote, which it will then display. The class is pretty simple, it only has the `getQuote` function that will modify the DOM, therefore it will be are main area of focus in testing.
+
+In order to test this component we need initiate the class `QuoteComponent`, the Angular testing library offers a way to inject any dependencies this class relies on, as well as a `TestComponentBuilder` object which will create the component for us and return a *component fixture* that we can perform tests on. 
+
+*quote.spec.ts*
+
+``` typescript
+class MockQuoteService {
+  public quote: "Test quote";
+  
+  getQuote() {
+  	return this.quote;
+  }
+}
+
+import {QuoteService} from 'quote.service.ts';
+import {QuoteComponent} from 'quote.component.ts';
+import {
+  expect, 
+  it, 
+  describe, 
+  injectAsync, 
+  TestComponentBuilder, 
+  beforeEachProvider
+} from 'angular2/testing';
+
+describe("Testing Quote Component", () => {
+    beforeEachProviders(() => {
+		provide(QuoteService: {useClass: MockQuoteService})
+	});
+    
+it("Should get quote", injectAsync([TestComponentBuilder], (tcb) => {
+  return tcb.createAsync(QuoteComponent).then(fixture) => {
+  		fixture.detectChanges();
```

Comment:
Do you know why is this first `detectChanges` needed? I understand the need of the second one that is trying to update the view with the new value of `this.quote`, but why the first one?